src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -53,6 +53,17 @@ void GcsNodeManager::HandleRegisterNode(rpc::RegisterNodeRequest request,
   RAY_LOG(INFO) << "Registering node info, node id = " << node_id
                 << ", address = " << request.node_info().node_manager_address()
                 << ", node name = " << request.node_info().node_name();
+  if (request.node_info().is_head_node()) {
+    // mark the old head node as dead
+    auto head_node_it =
+        std::find_if(alive_nodes_.begin(), alive_nodes_.end(), [](const auto &node) {
+          return node.second->is_head_node();
+        });
+    if (head_node_it != alive_nodes_.end()) {
+      NodeID old_head_node_id = head_node_it->first;
+      OnNodeFailure(old_head_node_id);
```

Comment:
OnNodeFailure is async: we need to make sure the old is marked as dead in redis before we add a new one.